[JSC] Add MayNeedDestruction mode

Constellation opened the pull request:
#### af0c5b6afa7e865b98f5900f8531e6ea2bc9b197
<pre>
[JSC] Add MayNeedDestruction mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=287318">https://bugs.webkit.org/show_bug.cgi?id=287318</a>
<a href="https://rdar.apple.com/144428259">rdar://144428259</a>

Reviewed by Yijia Huang.

This patch introduces MayNeedDestruction destruction mode. When this
mode is set, each cells need to manually track whether it should get
destruction. We use this in JSRopeString and we can mark MarkedBlock as
destructible only when we resolve JSRopeString contained in the
MarkedBlock. It is really common that we create massive amount of
JSRopeStrings and only a few gets resolved. In this case, most of
JSRopeStrings in many MarkedBlocks are not resolved. We can skip calling
destructors for these blocks.

Currently this mode is allowed only for types which only allocate cells
from MarkedBlocks. So JSRopeString is specifically made as it always
allocate cells from MarkedBlocks (not from PreciseAllocations).

* Source/JavaScriptCore/heap/BlockDirectory.cpp:
(JSC::BlockDirectory::endMarking):
* Source/JavaScriptCore/heap/BlockDirectory.h:
(JSC::BlockDirectory::attributes const):
(JSC::BlockDirectory::needsDestruction const): Deleted.
* Source/JavaScriptCore/heap/DestructionMode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/heap/DestructionMode.h:
(): Deleted.
* Source/JavaScriptCore/heap/HeapCell.h:
* Source/JavaScriptCore/heap/HeapCellInlines.h:
(JSC::HeapCell::notifyNeedsDestruction const):
* Source/JavaScriptCore/heap/IsoHeapCellType.h:
* Source/JavaScriptCore/heap/IsoInlinedHeapCellTypeInlines.h:
(JSC::IsoInlinedHeapCellType&lt;CellType&gt;::IsoInlinedHeapCellType):
* Source/JavaScriptCore/heap/MarkedBlock.cpp:
(JSC::MarkedBlock::Handle::stopAllocating):
(JSC::MarkedBlock::Handle::sweep):
* Source/JavaScriptCore/heap/MarkedBlock.h:
(JSC::MarkedBlock::Handle::needsDestruction const):
* Source/JavaScriptCore/heap/MarkedBlockInlines.h:
(JSC::MarkedBlock::Handle::specializedSweep):
(JSC::MarkedBlock::Handle::sweepDestructionMode):
(JSC::MarkedBlock::Handle::setIsDestructible):
* Source/JavaScriptCore/heap/PreciseAllocation.cpp:
(JSC::PreciseAllocation::sweep):
* Source/JavaScriptCore/runtime/ClonedArguments.h:
* Source/JavaScriptCore/runtime/DirectArguments.h:
* Source/JavaScriptCore/runtime/JSLexicalEnvironment.h:
(JSC::JSLexicalEnvironment::subspaceFor):
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSFinalObject::subspaceFor):
* Source/JavaScriptCore/runtime/JSString.h:
* Source/JavaScriptCore/runtime/JSStringInlines.h:
(JSC::JSRopeString::convertToNonRope const):
* Source/JavaScriptCore/runtime/RegExpObject.h:
* Source/JavaScriptCore/runtime/ScopedArguments.h:
* Source/JavaScriptCore/testRegExp.cpp:
* Source/WebCore/bindings/js/WebCoreJSClientData.h:
(WebCore::subspaceForImpl):

Canonical link: <a href="https://commits.webkit.org/290094@main">https://commits.webkit.org/290094@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7da4e1374b3c8d1648ae3ef9535972f85e273d75

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88928 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8452 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43524 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93899 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39687 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90979 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8839 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16636 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68513 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26188 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91930 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6735 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80505 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48877 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6484 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/34916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38796 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/81727 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76847 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35841 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95739 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/87704 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16108 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11754 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77390 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16364 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/76335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76678 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21062 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9174 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13936 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16122 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21433 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/110197 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15863 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26450 "Found 6 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.bytecode-cache, microbenchmarks/memcpy-wasm-medium.js.dfg-eager, microbenchmarks/memcpy-wasm-medium.js.no-llint, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager-jettison, wasm.yaml/wasm/stress/repro_1289.js.wasm-collect-continuously, wasm.yaml/wasm/stress/top-most-enclosing-stack.js.wasm-eager-jettison (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19314 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17644 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/heap/MarkedBlockInlines.h:
```diff
@@ -291,7 +291,7 @@ void MarkedBlock::Handle::specializedSweep(FreeList* freeList, MarkedBlock::Hand
         Locker locker { m_directory->bitvectorLock() };
         m_directory->setIsUnswept(this, false);
-        m_directory->setIsDestructible(this, false);
+        m_directory->setIsDestructible(this, destructionMode != BlockHasNoDestructors && !isEmpty && m_directory->isDestructible(this));
```
I see. I need to clear the bit once the block is cleared or it's refilled with objects w/o destructors. Nice!

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Add MayNeedDestruction mode
https://bugs.webkit.org/show_bug.cgi?id=287318
rdar://144428259

Reviewed by Yijia Huang.

This patch introduces MayNeedDestruction destruction mode. When this
mode is set, each cells need to manually track whether it should get
destruction. We use this in JSRopeString and we can mark MarkedBlock as
destructible only when we resolve JSRopeString contained in the
MarkedBlock. It is really common that we create massive amount of
JSRopeStrings and only a few gets resolved. In this case, most of
JSRopeStrings in many MarkedBlocks are not resolved. We can skip calling
destructors for these blocks.

Currently this mode is allowed only for types which only allocate cells
from MarkedBlocks. So JSRopeString is specifically made as it always
allocate cells from MarkedBlocks (not from PreciseAllocations).

* Source/JavaScriptCore/heap/BlockDirectory.cpp:
(JSC::BlockDirectory::endMarking):
* Source/JavaScriptCore/heap/BlockDirectory.h:
(JSC::BlockDirectory::attributes const):
(JSC::BlockDirectory::needsDestruction const): Deleted.
* Source/JavaScriptCore/heap/DestructionMode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/heap/DestructionMode.h:
(): Deleted.
* Source/JavaScriptCore/heap/HeapCell.h:
* Source/JavaScriptCore/heap/HeapCellInlines.h:
(JSC::HeapCell::notifyNeedsDestruction const):
* Source/JavaScriptCore/heap/IsoHeapCellType.h:
* Source/JavaScriptCore/heap/IsoInlinedHeapCellTypeInlines.h:
(JSC::IsoInlinedHeapCellType<CellType>::IsoInlinedHeapCellType):
* Source/JavaScriptCore/heap/MarkedBlock.cpp:
(JSC::MarkedBlock::Handle::stopAllocating):
(JSC::MarkedBlock::Handle::sweep):
* Source/JavaScriptCore/heap/MarkedBlock.h:
(JSC::MarkedBlock::Handle::needsDestruction const):
* Source/JavaScriptCore/heap/MarkedBlockInlines.h:
(JSC::MarkedBlock::Handle::specializedSweep):
(JSC::MarkedBlock::Handle::sweepDestructionMode):
(JSC::MarkedBlock::Handle::setIsDestructible):
* Source/JavaScriptCore/heap/PreciseAllocation.cpp:
(JSC::PreciseAllocation::sweep):
* Source/JavaScriptCore/runtime/ClonedArguments.h:
* Source/JavaScriptCore/runtime/DirectArguments.h:
* Source/JavaScriptCore/runtime/JSLexicalEnvironment.h:
(JSC::JSLexicalEnvironment::subspaceFor):
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSFinalObject::subspaceFor):
* Source/JavaScriptCore/runtime/JSString.h:
* Source/JavaScriptCore/runtime/JSStringInlines.h:
(JSC::JSRopeString::convertToNonRope const):
* Source/JavaScriptCore/runtime/RegExpObject.h:
* Source/JavaScriptCore/runtime/ScopedArguments.h:
* Source/JavaScriptCore/testRegExp.cpp:
* Source/WebCore/bindings/js/WebCoreJSClientData.h:
(WebCore::subspaceForImpl):

Canonical link: https://commits.webkit.org/290094@main
